crates/block-producer/src/block_builder/prover/block_witness.rs

```diff
@@ -77,13 +48,83 @@ impl BlockWitness {
             .map(|(batch_index, batch)| (batch_index, batch.output_notes_root()))
             .collect();
 
-        Ok(Self {
-            updated_accounts,
-            batch_created_notes_roots,
-            produced_nullifiers: block_inputs.nullifiers,
-            chain_peaks: block_inputs.chain_peaks,
-            prev_header: block_inputs.block_header,
-        })
+        // Order account updates by account ID and each update's initial state hash.
+        //
+        // This let's us chronologically order the updates per account across batches.
+        let mut updated_accounts = BTreeMap::<AccountId, BTreeMap<Digest, AccountUpdate>>::new();
+        for (account_id, update) in batches.iter().flat_map(TransactionBatch::updated_accounts) {
+            updated_accounts
+                .entry(*account_id)
+                .or_default()
+                .insert(update.init_state, update.clone());
+        }
+
+        // Build account witnesses.
+        let mut account_witnesses = Vec::with_capacity(updated_accounts.len());
+        let mut block_updates = Vec::with_capacity(updated_accounts.len());
+
+        for (account_id, mut updates) in updated_accounts {
+            let (initial_state_hash, proof) = block_inputs
+                .accounts
+                .remove(&account_id)
+                .map(|witness| (witness.hash, witness.proof))
+                .ok_or(BuildBlockError::InconsistentAccountIds(vec![account_id]))?;
+
+            let mut details: Option<AccountUpdateDetails> = None;
+
+            // Chronologically chain updates for this account together using the state hashes to link them.
+            let mut transactions = Vec::new();
+            let mut current_hash = initial_state_hash;
+            while !updates.is_empty() {
+                let update = updates
+                    .remove(&current_hash)
+                    .ok_or(BuildBlockError::InconsistentAccountStates(vec![account_id]))?;
+
+                transactions.extend(update.transactions);
+                current_hash = update.final_state;
+
+                details = Some(match details {
+                    None => update.details,
+                    Some(details) => {
+                        details.merge(update.details).expect("Should this be an error?")
```

Comment:
I wasn't sure if detail merging could fail at this stage? If yes, what error does this become? Till now only store and batch updates could mismatch, now batches can mismatch as well..